stan/math/rev/functor/algebra_solver_powell.hpp

```diff
@@ -21,57 +22,77 @@ namespace math {
  * function theorem. The call to Jacobian() occurs outside the call to
  * chain() -- this prevents malloc issues.
  */
-template <typename Fs, typename F, typename T, typename Fx>
+template <typename Fy, typename T, typename Fx>
 struct algebra_solver_vari : public vari {
-  /** vector of parameters */
-  vari** y_;
   /** number of parameters */
-  int y_size_;
+  const int y_size_;
+  /** value of parameters */
+  const Eigen::Matrix<T, Eigen::Dynamic, 1> y_val_;
+  /** System functor (f_) w.r.t. inputs (y_) */
+  Fy fy_;
+  /** array of parameters */
+  vari** y_;
   /** number of unknowns */
-  int x_size_;
-  /** vector of solution */
-  vari** theta_;
-  /** Jacobian of the solution w.r.t parameters */
-  double* Jx_y_;
+  const int x_size_;
+  /** value of unknowns */
+  const Eigen::VectorXd& x_val_;
+  /** Hybrj functor solver (f_) w.r.t. outputs (x_) */
+  Fx fx_;
+  /** array of unknowns */
+  vari** x_;
+  /** Jacobian of f w.r.t. outputs (x_) */
+  const Eigen::MatrixXd Jf_x_;
 
-  algebra_solver_vari(const Fs& fs, const F& f, const Eigen::VectorXd& x,
-                      const Eigen::Matrix<T, Eigen::Dynamic, 1>& y,
-                      const std::vector<double>& dat,
-                      const std::vector<int>& dat_int,
-                      const Eigen::VectorXd& theta_dbl, Fx& fx,
-                      std::ostream* msgs)
-      : vari(theta_dbl(0)),
-        y_(ChainableStack::instance_->memalloc_.alloc_array<vari*>(y.size())),
+  algebra_solver_vari(Fy& fy, const Eigen::Matrix<T, Eigen::Dynamic, 1>& y,
+                      Fx& fx, const Eigen::VectorXd& x)
+      : vari(x(0)),
         y_size_(y.size()),
+        y_val_(y),
+        fy_(fy),
+        y_(ChainableStack::instance_->memalloc_.alloc_array<vari*>(y_size_)),
         x_size_(x.size()),
-        theta_(
-            ChainableStack::instance_->memalloc_.alloc_array<vari*>(x_size_)),
-        Jx_y_(ChainableStack::instance_->memalloc_.alloc_array<double>(
-            x_size_ * y_size_)) {
+        x_val_(x),
+        Jf_x_(fx.get_jacobian(x_val_)),
+        fx_(fx),
+        x_(ChainableStack::instance_->memalloc_.alloc_array<vari*>(x_size_)) {
```

Comment:
Is `X` from where this is called a matrix of vars? I think you could just do `arena_t<Eigen::Matrix<var, -1, 1>> x_` as a member and then access it's values and adjoints via `x_.val()` and `x_.adj()`

Reply:
I think it's actually not a matrix of `var`s—the Powell / Newton solvers return a vector of doubles, and then this is the first place where those doubles are being packaged in `var`/`vari`s. Although, if a vector of doubles can be implicitly converted to a vector of `var`, maybe it would be better to do something like this?